taichi/ui/backends/vulkan/renderable.cpp

```diff
@@ -0,0 +1,301 @@
+      }
+      update_renderables_vertices_cuda(
+          vertex_buffer_device_ptr_, (float *)info.per_vertex_color.data,
+          num_vertices, 3, offsetof(Vertex, color));
+    }
+
+    if (info.normals.valid) {
+      if (info.normals.shape[0] != num_vertices) {
+        throw std::runtime_error(
+            "shape of normals should be the same as vertices");
+      }
+      update_renderables_vertices_cuda(vertex_buffer_device_ptr_,
+                                       (float *)info.normals.data, num_vertices,
+                                       3, offsetof(Vertex, normal));
+    }
+
+    if (info.indices.valid) {
+      indexed_ = true;
+      update_renderables_indices_cuda(index_buffer_device_ptr_,
+                                      (int *)info.indices.data, num_indices);
+    } else {
+      indexed_ = false;
+    }
+
+  } else if (info.vertices.field_source == FieldSource::TaichiX64) {
+    {
+      Vertex *mapped_vbo =
+          (Vertex *)app_context_->device().map(staging_vertex_buffer_);
+
+      update_renderables_vertices_x64(mapped_vbo, (float *)info.vertices.data,
+                                      num_vertices, num_components,
+                                      offsetof(Vertex, pos));
+      if (info.per_vertex_color.valid) {
+        if (info.per_vertex_color.shape[0] != num_vertices) {
+          throw std::runtime_error(
+              "shape of per_vertex_color should be the same as vertices");
+        }
+        update_renderables_vertices_x64(
+            mapped_vbo, (float *)info.per_vertex_color.data, num_vertices, 3,
+            offsetof(Vertex, color));
+      }
+      if (info.normals.valid) {
+        if (info.normals.shape[0] != num_vertices) {
+          throw std::runtime_error(
+              "shape of normals should be the same as vertices");
+        }
+        update_renderables_vertices_x64(mapped_vbo, (float *)info.normals.data,
+                                        num_vertices, 3,
+                                        offsetof(Vertex, normal));
+      }
+      app_context_->device().unmap(staging_vertex_buffer_);
+
+      int *mapped_ibo =
+          (int *)app_context_->device().map(staging_index_buffer_);
+      if (info.indices.valid) {
+        indexed_ = true;
+        update_renderables_indices_x64(mapped_ibo, (int *)info.indices.data,
+                                       num_indices);
+      } else {
+        indexed_ = false;
+      }
+      app_context_->device().unmap(staging_index_buffer_);
+    }
+    app_context_->device().memcpy(vertex_buffer_.get_ptr(0),
+                                  staging_vertex_buffer_.get_ptr(0),
+                                  config_.vertices_count * sizeof(Vertex));
+    app_context_->device().memcpy(index_buffer_.get_ptr(0),
+                                  staging_index_buffer_.get_ptr(0),
+                                  config_.indices_count * sizeof(int));
```

Comment:
btw when needing more than 1 copy, we should get a new command list and record two copy commands. The device memcpy call is synced and we should try to reduce the number of command lists submitted

Reply:
good point, will fix.